sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -424,11 +424,19 @@ case class FileSourceScanExec(
     val defaultMaxSplitBytes =
       fsRelation.sparkSession.sessionState.conf.filesMaxPartitionBytes
     val openCostInBytes = fsRelation.sparkSession.sessionState.conf.filesOpenCostInBytes
-    val defaultParallelism = fsRelation.sparkSession.sparkContext.defaultParallelism
-    val totalBytes = selectedPartitions.flatMap(_.files.map(_.getLen + openCostInBytes)).sum
-    val bytesPerCore = totalBytes / defaultParallelism
 
-    val maxSplitBytes = Math.min(defaultMaxSplitBytes, Math.max(openCostInBytes, bytesPerCore))
+    // Ignore bytesPerCore when dynamic allocation is enabled. See SPARK-22411
+    val maxSplitBytes =
+      if (Utils.isDynamicAllocationEnabled(fsRelation.sparkSession.sparkContext.getConf)) {
+        defaultMaxSplitBytes
```

Comment:
does it really help? Without this change, `bytesPerCore` is usually very big and `maxSplitBytes` will be just `defaultMaxSplitBytes`

Reply:
That is true. In that case, this change wouldn't help. That is why I asked if using `spark.dynamicAllocation.maxExecutors` for calculating bytesPerCore is better when dynamic allocation is enabled. 

For small tables (when `bytesPerCore` is less than `defaultMaxSplitBytes`) we have had use cases where we wanted control over the partition size without being limited by `bytesPerCore`. We internally added a flag to ignore `bytesPerCore`. 

Reasoning about split size in terms of the two parameters 'defaultMaxSplitBytes' and 'openCostInBytes' seems to be more intuitive. What do you think?

Reply:
cc @jerryshao

Reply:
What if `spark.dynamicAllocation.maxExecutors` is not configured? Seems we cannot rely on this configuration, user may not always set it.

My concern is the cost of ramping up new executors, by splitting partitions into small ones, Spark will ramp up more executors to execute small tasks, when the cost of ramping up new executors is larger than executing tasks, this seems not a heuristic solution anymore. Previously because all the executors are available, so heuristic solution is valid.

For small data (calculated `bytesPerCore ` < `defaultMaxSplitBytes `, less than 128M), I think using the available resources to schedule tasks would be enough, since task is not so big. For big data (calculated `bytesPerCore ` > `defaultMaxSplitBytes `, larger than 128M), I think 128M might be the proper value to issue new executors and tasks. So IMHO seems current solution is sufficient for dynamic allocation scenario. Please correct me if I'm wrong.

Reply:
@jerryshao That's true. We cannot rely on users to set it.

For small data, sometimes users care about the number of output files generated as well. To give the flexibility to the users, how about adding a flag to ignore `bytesPerCore` when set regardless of dynamic allocation being used or not? That way the default behavior stays as is and users with specific needs can set the flag.

Reply:
>For small data, sometimes users care about the number of output files generated as well.

Can you please elaborate more, do you want more output files or less output files. if less, I think I already mentioned in the above comment.

Seems you patch already ignored `bytesPerCore` when dynamic allocation is enabled, are you suggesting something different?

Reply:
@jerryshao When `bytesPerCore` is less than `defaultMaxSplitBytes` and  is chosen as the `maxSplitsBytes` instead of `defaultMaxSplitBytes`, the number of output files is greater than what it would be when `maxSplitBytes`=`defaultMaxSplitBytes`. So for use cases against small data where users want fewer output files  they should have the flexibility to do so.

Instead of ignoring `bytesPerCore` when dynamic allocation is enabled I am proposing adding a new flag `overrideMaxPartitionSize`. When this is set, `maxSplitBytes` would be equal to `defaultMaxSplitBytes` (i.e, the current heuristic will not be used). The default value of this flag will be false, so that the current behavior is not affected.